src/js/GoldifyExecutePage.js

```diff
@@ -1,14 +1,196 @@
 import "../css/GoldifyExecutePage.css";
 
+const queryString = require("query-string");
+const goldifyUtils = require("../utils/GoldifyUtils");
+const goldifyTestUtils = require("../utils/GoldifyTestUtils");
+
+var clientId = process.env.REACT_APP_SPOTIFY_CLIENT_ID; // Your client id
+var clientSecret = process.env.REACT_APP_SPOTIFY_CLIENT_SECRET; // Your secret
+var redirectUri = process.env.REACT_APP_SPOTIFY_REDIRECT_URI; // Your redirect uri
+
+/**
+ * Generates a random string containing numbers and letters
+ * @param  {number} length The length of the string
+ * @return {string} The generated string
+ */
+var generateRandomString = function(length) {
+  var text = '';
+  var possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
+  for (var i = 0; i < length; i++) {
+    text += possible.charAt(Math.floor(Math.random() * possible.length));
+  }
+  return text;
+};
+
+function useQuery() {
+  return new URLSearchParams(window.location.search);
+}
+
+function retrieveAuthorization() {
+  let spotifyApiURL =
+    "https://accounts.spotify.com/authorize?" +
+    queryString.stringify({
+      response_type: "code",
+      client_id: clientId,
+      scope: "user-read-private user-read-email",
+      redirect_uri: redirectUri,
+      state: generateRandomString(16)
+    });
+  replaceWindowURL(spotifyApiURL);
+}
+
+function replaceWindowURL(url) {
+  if (!goldifyUtils.jestTestIsRunning()) {
+    window.location.replace(url);
+  }
+}
+
 class GoldifyExecutePage extends Component {
-  render() {
+  
+  constructor(props) {
+    // Initialize mutable state
+    super(props);
+    this.state = {
+      initFailed: false,
+      authorizationCode: "",
+      refreshToken: "",
+      accessToken: "",
+      userData: null,
+      userDataString: ""
+    };
+  }
+
+  componentDidMount() {
+    let code = useQuery().get('code');
+    if (code == undefined || code == null) {
+      retrieveAuthorization();
+      if (goldifyUtils.jestTestIsRunning()) {
+        this.retrieveTokens("TEST_AUTH_CODE");
+      }
+    } else {
+      this.retrieveTokens(code);
+    }
+  }
+
+  retrieveTokens(code) {
+    let apiTokenBody = new URLSearchParams();
+    apiTokenBody.set("grant_type", "authorization_code");
+    apiTokenBody.set("redirect_uri", redirectUri);
+    apiTokenBody.set("code", code);
+    let spotifyApiTokenURL = "https://accounts.spotify.com/api/token";
+    fetch(spotifyApiTokenURL, {
+      method: "POST",
+      headers: {
+        "Authorization": " Basic " + btoa(clientId + ":" + clientSecret)
+      },
+      body: apiTokenBody
+    })
+      .then((response) => {
+        if (!response.ok) {
+          replaceWindowURL("/");
+        }
+        return response.json();
+      })
+      .then((data) => {
+        this.retrieveUserData(code, data);
+      })
+      .catch(error => console.log(error)); // eslint-disable-line no-console
+    if (goldifyUtils.jestTestIsRunning()) {
+      this.setStateForTest();
```

Comment:
Same here, if we could just have lots of unit tests and mocked function calls instead, we should be able to remove all logic only related to testing

Reply:
Same as comment above

```
This is a great point, and not the right move by me. Could get very sloppy.

For record keeping, this has been resolved by using mocking in our test suites.
```